[WIP] v2.2

elmigranto opened the pull request:
v2.2 milestone with legacy data convertion <s>(branched off v2.1)</s>.

elmigranto committed:
initial legacy data convertion

elmigranto committed:
fixed "supervision" field type

elmigranto committed:
converting legacy procedure.picture to procedure.photos

elmigranto committed:
fixed editableView's discard for templateView

elmigranto committed:
default template with fields from v1

elmigranto committed:
remove safekeep, keep original item instead, but remember we converted it

elmigranto committed:
updated camera code

elmigranto commented:
Almost finished, will try to build cordova app, can you give it a try too, because I am not sure it will build again…

elmigranto committed:
fixed typo

elmigranto committed:
fixed filed restoration leading to creating new fields

j3k0 commented:
Trying it up.

First notice: `npm test` fails.

```
verity:logbookr jeko$ npm test

> logbookr@0.1.0 test /Users/jeko/Documents/logbookr
> mocha mocha-tests



  LocalStorage
    converters
      .startsWith
        ✓ returns true if string starts with beginning
        ✓ returns false otherwise
      .shouldConvert()
        1) returns true on keys that are not prefixed with `config.dataPrefix`
        2) returns false otherwise
      .convert()
        returns an object
          ✓ has only requried keys
          ✓ all values are strings, null is allowed as item
          ✓ `key` is dataPrefix + key
        correctly converts…
          ✓ ChTr
          ✓ ChTr-{id}
          ✓ Procs
          3) Procs-{id}
          ✓ returns null as key/item for other types of keys

  FieldModel
    ✓ exports available field types
    ✓ has `name`, `type`, `description`
    ✓ `type` defaults to TEXT
    #validate()
      ✓ `type` must be one of FieldModel.types
      ✓ `name` must be non-empty string
      ✓ returns undefined on well-formed fields
    #defaultValue()
      ✓ empty array for PHOTOS
      ✓ empty string for other types

  ProcedureModel
    defaults() / initialize()
      ✓ defines meta info: `id`, `createdAt`, `requiredFields` and `fields`
      ✓ each required field has an according model attribute
      ✓ when no options.template is specified, attribute `fields` is an emtpy array
      ✓ when options.template is specified, copies over fields from options.template.get('fields') to `fields` attribute
      #clone()
        ✓ returns new instance with deep cloned attributes
      #diff()
        ✓ returns set of changes required to revert model to its previous state
        ✓ returns `false` when model is unchagned compared to its previous state
      #revert()
        ✓ returns model to its previous state (before latest `set()`)
        ✓ does nothing if model was not changed
      #safeSet()
        ✓ applies changes and returns true
        ✓ preserves model.validationError
        ✓ returns false and reverts changes if, when applied, resulted in invalid model
        ✓ throws Error if changes argument is incorrect
      #fieldInfo()
        ✓ returns field info by its name
        ✓ returns undefined if fields was not found

  tr()
    ✓ returns `en` translations
    ✓ returns `fr` translations
    ✓ warns about missing locales
    ✓ warns about missing keys


  36 passing (125ms)
  3 failing

  1) LocalStorage converters .shouldConvert() returns true on keys that are not prefixed with `config.dataPrefix`:
     TypeError: undefined is not a function
      at shouldConvert (src/local-storage/converters.js:21:21)
      at Context.<anonymous> (mocha-tests/local-storage.test.js:24:16)

  2) LocalStorage converters .shouldConvert() returns false otherwise:
     TypeError: undefined is not a function
      at shouldConvert (src/local-storage/converters.js:21:21)
      at Context.<anonymous> (mocha-tests/local-storage.test.js:29:16)

  3) LocalStorage converters .convert() correctly converts… Procs-{id}:

      Error: expected { id: '30b241d9-3294-1e9d-f193-8385a9e34afb',
  createdAt: 1445273520000,
  date: '19-10-2015 19:52',
  type: 'Discectomie simple',
  patient: 'deep tree',
  photos: [],
  diagnostic: '',
  supervision: 'Aidé',
  senior: '',
  stage: 'c',
  comment: '',
  fields:
   [ { name: 'diagnostic',
       description: 'Diagnostic',
       type: 'text' },
     { name: 'supervision',
       description: 'Supervision',
       type: 'choicetree' },
     { name: 'senior',
       description: 'Senior',
       type: 'choicetree' },
     { name: 'stage',
       description: 'Stage',
       type: 'choicetree' },
     { name: 'comment',
       description: 'Comment',
       type: 'textarea' } ],
  requiredFields:
   [ { name: 'date',
       description: 'Date Performed',
       type: 'date' },
     { name: 'type',
       description: 'Type of Procedure',
       type: 'choicetree' },
     { name: 'patient',
       description: 'Patient\'s Full Name',
       type: 'text' },
     { name: 'photos',
       description: 'Photos',
       type: 'photos' } ] } to sort of equal { id: '30b241d9-3294-1e9d-f193-8385a9e34afb',
  createdAt: 1445266320000,
  type: 'Discectomie simple',
  date: '19-10-2015 19:52',
  patient: 'deep tree',
  photos: [],
  requiredFields:
   [ { name: 'date',
       description: 'Date Performed',
       type: 'date' },
     { name: 'type',
       description: 'Type of Procedure',
       type: 'choicetree' },
     { name: 'patient',
       description: 'Patient\'s Full Name',
       type: 'text' },
     { name: 'photos',
       description: 'Photos',
       type: 'photos' } ],
  diagnostic: '',
  supervision: 'Aidé',
  senior: '',
  stage: 'c',
  comment: '',
  fields:
   [ { description: 'Diagnostic',
       name: 'diagnostic',
       type: 'text' },
     { description: 'Supervision',
       name: 'supervision',
       type: 'choicetree' },
     { description: 'Senior',
       name: 'senior',
       type: 'choicetree' },
     { description: 'Stage',
       name: 'stage',
       type: 'choicetree' },
     { description: 'Comment',
       name: 'comment',
       type: 'textarea' } ] }
      + expected - actual

       {
         "comment": ""
      -  "createdAt": 1445273520000
      +  "createdAt": 1445266320000
         "date": "19-10-2015 19:52"
         "diagnostic": ""
         "fields": [
           {

      at Assertion.assert (node_modules/expect.js/index.js:96:13)
      at Assertion.eql (node_modules/expect.js/index.js:230:10)
      at mocha-tests/local-storage.test.js:80:51
      at Array.forEach (native)
      at Context.<anonymous> (mocha-tests/local-storage.test.js:78:28)



npm ERR! Test failed.  See above for more details.
```

elmigranto committed:
fixed typo

elmigranto committed:
"fixed" timezones in convertion test.

elmigranto commented:
Interesting…

First two failing is because of a typo — used `key.startsWith(…)` instead of `startsWith(key…)`, but my node is 4.2.1 and there is `String.prototype.startsWith` in it. Fixed that.

The other one is a bit trickier. So since there is no information about timezone in a date, it parses into 2 different timestamps on mine PC and yours, and I used timestamp in mine tz during tests. Never occurred to me, changed that to be calculated before tests run.
